Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1064 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import DequeModule
+import NIOConcurrencyHelpers
+
+#if compiler(>=5.5.2) && canImport(_Concurrency)
+/// A protocol for the back-pressure strategy of the ``NIOAsyncSequenceProducer``.
+///
+/// A back-pressure strategy is invoked when new elements are yielded to the sequence or
+/// when a ``NIOAsyncSequenceProducer/AsyncIterator`` requested the next value. The responsibility of the strategy is
+/// to determine whether more elements need to be produced .
+///
+/// If more elements need to be produced, either the ``NIOAsyncSequenceProducerDelegate/produceMore()``
+/// method will be called or a ``NIOAsyncSequenceProducer/Source/YieldResult`` will be returned that indicates
+/// to produce more.
+///
+/// - Important: The methods of this protocol are guaranteed to be called serially. Furthermore, the implementation of these
+/// methods **MUST NOT** do any locking or call out to any other Task/Thread.
+public protocol NIOAsyncSequenceProducerBackPressureStrategy: Sendable {
```

Comment:
I think for most of the cases the two strategies that we are going to implement are used.

Reply:
Do we absolutely want this to be generic then? 

Just wondering whether it would make sense instead to provide an enum-backed-struct with those two strategies and add more (or an escape hatch) if required. As a side effect it would make it mildly easier to construct the async source/sequence and avoid the naming issues.

Reply:
Open to do this change. Would like to get @Lukasa opinion on this as well, as it would basically mean we need to implement all strategies inside NIO.

Reply:
I'm inclined to keep the generic: specialisation should allow the strategies to more-or-less go away. It does make the type hard to spell, I agree, but we don't want users to expose this type in their API regardless.